BV2-7 단일 이미지 저장 구현

seongeun308 opened the pull request:
## #️⃣ Issue Number

<!--- ex) #이슈번호, #이슈번호 -->

## 📝 요약(Summary)

<!--- 변경 사항 및 관련 이슈에 대해 간단하게 작성해주세요. 어떻게보다 무엇을 왜 수정했는지 설명해주세요. -->

## 🛠️ PR 유형

어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [x] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## 💬 공유사항 to 리뷰어

<!--- 리뷰어가 중점적으로 봐줬으면 좋겠는 부분이 있으면 적어주세요. -->
<!--- 논의해야할 부분이 있다면 적어주세요.-->
<!--- ex) 메서드 XXX의 이름을 더 잘 짓고 싶은데 혹시 좋은 명칭이 있을까요? -->

## 📑 API Spec

| **기능**          |                |
|-----------------|----------------|
| **HTTP Method** |                |
| **URL**         |                |

### Request Body

### Response Body

## 📸스크린샷 (선택)


## ✅ PR Checklist

PR이 다음 요구 사항을 충족하는지 확인하세요.

- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다. (버그 수정/기능에 대한 테스트)

seongeun308 committed:
BV2-7 모듈 통신 및 이미지 저장 추상화

seongeun308 committed:
BV2-7 (feat): 이미지 저장을 위한 로컬파일스토리지 및 MongoDB 구현

MongoDB에 저장할 ImageMetaData 엔티티 추가
MongoDB 의존성 추가
MongoDB 설정 추가

seongeun308 committed:
BV2-7 (test): 이미지 저장을 위한 로컬파일스토리지 및 MongoDB 테스트

파일스토리지 테스트를 위해 Mockito 사용

seongeun308 committed:
BV2-7 (fix): ImageProducer send() 매개변수 수정

ImageResponse responses -> String topic, ImageResponse response

seongeun308 committed:
BV2-7 (refactor): 이미지 저장 시 필요한 정보를 위한 FileUtils 추가 및 적용

seongeun308 committed:
BV2-7 (feat): 파일스토리지에 이미지 저장 후 메타데이터를 DB에 저장하는 LocalFileAndDBImageStorageService 구현

파일스토리지 저장 후 serverName 반환
serverName로 ImageMetaData 생성
ImageMetaData 저장

seongeun308 committed:
BV2-7 (feat): kafka를 통한 단일 이미지 저장 요청과 응답 처리

shinywoon commented on file/src/main/java/file/application/port/input/ImageConsumer.java:
```diff
@@ -0,0 +1,9 @@
+
+public interface ImageConsumer {
+    void consume(ImageRequest imageRequest);
+}
```
Produce, Consumer 부분은 `global` Module 내 작성된 Message Producer Interface 를 이용하시면 됩니다. 
```JAVA
public interface MessageProducer<T> {
    void send(String topic, T message);
}
```
```JAVA
public interface MessageConsumer<T> {
    T consumeMessage(T message);
}
```
Generic 으로 작성되어 있어 사용하고자하는 위치에서 T 위치에 자신이 이용할 객체를 선언하시면 됩니다.
[example]
```java
public class KafkaProducer implements MessageProducer<ImageMessage> {
  ... 
}
```
`global` Module 내 객체를 사용하기 위해서는 `build.gradle` file 에 dependencies 범위 내 `implementation project(":global")` 선언 하시면 됩니다.
(example)
```groovy
dependencies {
  implementation project(":global")
}
```

shinywoon commented on file/src/main/java/file/application/port/input/ImageStorageService.java:
```diff
@@ -0,0 +1,8 @@
+
+public interface ImageStorageService {
+    ImageMetaData saveImage(ImageRequest imageRequest);
+}
```
ImageStorageService 보다는 `ImageStorageUsecase` 명칭이 좀더 명확합니다.
해당 interface를 implements 하는 `service` 객체와 구분하고자 합니다.
interface 만 봐서는 input port 인지 component 인지 구별이 되지 않습니다.
좀 더 명확한 표현이 필요합니다.

shinywoon commented on file/src/main/java/file/domain/ImageRequest.java:
```diff
@@ -0,0 +1,4 @@
+package file.domain;
+
+public class ImageRequest {
+}
```
객체 명칭이 좀 더 명확하게 작성되었으면 좋겠습니다.
또, 헥사고날에서는 들어오는 입력 요청을 Command라고 지칭합니다.

shinywoon commented on file/src/main/resources/application.yml:
```diff
@@ -0,0 +1,22 @@
+server:
+  port: 8083
```
Port 설정은 random 설정을 위해 0 처리 하였습니다.
upload-dir 부분은 타 개발자들의 환경을 위하여 상대 경로 처리 되거나 Classpath 처리 되어야 합니다.

shinywoon commented on file/src/main/resources/application.yml:
```diff
@@ -0,0 +1,22 @@
+    producer:
+      bootstrap-servers: localhost:9092
+      key-serializer: org.apache.kafka.common.serialization.StringSerializer
+      value-serializer: org.apache.kafka.common.serialization.StringSerializer
```
제가 올린 설정 파일 참조해주세요!

shinywoon commented on file/src/main/java/file/adapter/output/LocalFileStorage.java:
```diff
@@ -0,0 +1,48 @@
+        log.info("Uploaded image to {}", filePath);
+
+        return serverName;
+    }
```
1. Method 의 complexity 가 높습니다. 추상화 레벨을 조정하여 complexity를 낮추고 가독성을 높혀주세요.

shinywoon commented on file/src/main/java/file/adapter/output/exception/FileStorageException.java:
```diff
@@ -0,0 +1,7 @@
+public class FileStorageException extends RuntimeException{
+    public FileStorageException(String message) {
+        super(message);
+    }
```
내부 Exception 구현 및 Handling 기준이 존재합니다.
내부 기준에 따라 작성해주세요!

shinywoon reviewed (changes requested):
프로젝트에 대해 설명을 해줄 시간이 필요할 것 같은데
한번 모여서 이야기하면서 기준에 대해 말씀드릴께요.

seongeun308 committed:
Merge remote-tracking branch 'origin/dev' into feat/file

seongeun308 committed:
BV2-7 (refactor):  PR 리뷰사항 적용

- global Module 내 작성된 Message Producer Interface 를 이용
- 명확성을 위해 ImageStorageService -> ImageStorageUsecase 명칭 변경
- 명확성을 위해 ImageRequest -> ImageCommand 명칭 변경
- 설정 파일 변경
- 내부 Exception 구현 재사용

seongeun308 committed:
BV2-7 (fix): Kafka 제거

kafka를 통한 모듈 통신이 아닌 패키지를 가져와 직접 사용하는 방식으로 변경

seongeun308 committed:
BV2-7 (feat): 서비스 비동기 처리 및 서비스 분리

- LocalFileStorageService, MongDBImageMetaDataService 추가
- ImageMetaData extension 속성 추가
- FileUtils을 통해 파일의 이름과 확장자를 가져오는 메서드 추가

shinywoon committed:
Merge branch 'dev' into feat/file
